scanner/src/test/kotlin/scanners/fossid/FossIdTest.kt

```diff
@@ -0,0 +1,829 @@
+private const val SCAN_ID = "testScanId"
+
+/**
+ * Create a new [FossId] instance with the specified [config].
+ */
+private fun createFossId(config: FossIdConfig): FossId =
+    FossId("fossId", ScannerConfiguration(), DownloaderConfiguration(), config)
+
+/**
+ * Create a standard [FossIdConfig] whose properties can be partly specified.
+ */
+private fun createConfig(
+    waitForResult: Boolean = true,
+    packageNamespaceFilter: String = "",
+    packageAuthorsFilter: String = "",
+    deltaScans: Boolean = true,
+    loginPage: String = LOGIN_PAGE
+): FossIdConfig {
+    val config = FossIdConfig(
+        "https://www.example.org/fossid",
+        API_KEY, USER, waitForResult, packageNamespaceFilter, packageAuthorsFilter, addAuthenticationToUrl = false,
+        deltaScans, emptyMap()
+    )
+
+    val service = createServiceMock(loginPage)
+    val namingProvider = createNamingProviderMock()
+    val configSpy = spyk(config)
+    every { configSpy.createService() } returns service
+    every { configSpy.createNamingProvider() } returns namingProvider
+
+    return configSpy
+}
+
+/**
+ * Create a mock [FossIdNamingProvider] that returns deterministic names derived from the parameters provided to its
+ * _createXXX()_ functions.
+ */
+private fun createNamingProviderMock(): FossIdNamingProvider {
+    val provider = mockk<FossIdNamingProvider>()
+
+    every { provider.createProjectCode(any()) } answers {
+        projectCode(firstArg())
+    }
+
+    every { provider.createScanCode(any(), any()) } answers {
+        scanCode(firstArg(), secondArg())
+    }
+
+    return provider
+}
+
+/**
+ * Create a mock for the [FossIdRestService]. The mock is prepared to return the given [loginPage] content. (This is
+ * queried directly in the constructor of [FossId].)
+ */
+private fun createServiceMock(loginPage: String): FossIdRestService {
+    val service = mockk<FossIdRestService>()
+
+    coEvery { service.getLoginPage() } returns loginPage.toResponseBody()
+
+    return service
+}
+
+/**
+ * Generate a project code for the project with the given [name].
+ */
+private fun projectCode(name: String): String = "$name:projectCode"
+
+/**
+ * Generate a synthetic scan code for the project with the given [name] and [tag].
+ */
+private fun scanCode(name: String, tag: FossId.DeltaTag?): String = "$name:${tag?.name}:scanCode"
+
+/**
+ * Create a mock [ScanStatus] that returns the given [state].
+ */
+private fun createScanStatus(state: ScanState): ScanStatus {
+    val status = mockk<ScanStatus>()
+    every { status.state } returns state
+    every { status.comment } returns "status$state"
+    return status
+}
+
+/**
+ * Create a mock [Scan] with the given properties.
+ */
+private fun createScan(url: String, revision: String, scanCode: String): Scan {
+    val scan = mockk<Scan>()
+    every { scan.gitRepoUrl } returns url
+    every { scan.gitBranch } returns revision
+    every { scan.code } returns scanCode
+    return scan
```

Comment:
For my knowledge: what is the difference between creating such a mock with `every` and constructing a plain object as in next function ?

Reply:
`Scan` has a bunch of properties. I was just to lazy to initialize them all.